### DISCUSSION TITLE
Release v1.0.0

### DISCUSSION THREAD

**Discussion by agentgt** [jstachio/jstachio#98] [Polls]
@dsyer @sviperll 

I would like to release 1.0.0.

I think it is finally ready to seal the API. Once 1.0.0 is released changes will be done through a more PR model including even myself. 

Please kick the tires if you can particularly the documentation. I'm sure there are typos but one I'm most concerned with is glaring API issues.

The documentation is here:
https://jstach.io/jstachio/

It's entirely Javadoc driven.

Regardless thank you guys for all the help.

**Comment by sviperll** [comment]
I would probably vote yes, but since you are pretty serious about 1.0 guarantees, I've made a quick pass through API and here are some unsorted comments that I've got after reading.

 * Maybe we should rename `SpecFormatter` to `NullToleratingFormatter` so it is much more obvious why it is not the default.
 * I think `Formatter` and `Escaper` interfaces should be annotated with `@FunctionalInterface`
 * I think `Formatter` and `Escaper` interfaces should be used everywhere instead of `Function<Object, String>` or `Function<String, String>`. This will make API definition shorter and more clear and this should provide the same functionality if `Formatter` and `Escaper` are functional interfaces.
 * [`TemplateInfo::lastLoaded`](https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/TemplateInfo.html#lastLoaded()) uses long as a timestamp and this should probably be `java.time.Instant`.

### More ###

#### JStashioFilter ####

I'm not really a user of jstachio, but from reading the docs I'm not sure what exactly `JStashioFilter` does and what `JStachioFilter.FilterChain` is. Should this really be a public API? It looks more like an implementation detail for me...

#### Escaper::apply ####

I'm not sure what [`Escaper::apply`](https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/Escaper.html#apply) method does. it says

 > Escapes a String by using StringBuilder and calling [Appender.append(Appendable, CharSequence)](https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/Appender.html#append(A,java.lang.CharSequence)).
>
> Throws: [UncheckedIOException](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/UncheckedIOException.html) - if the appender or appendable throw an [IOException](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/IOException.html)

But I'm not sure where the `Appender` comes into the picture. My understanding was that Escaper is just a pure function to escape strings... Are these implementation details that should not leak into a public API definition?

**Comment by agentgt** [comment]
> I think Formatter and Escaper interfaces should be annotated with @FunctionalInterface

I always forget to annotate functional interfaces. I will add that.

>  think Formatter and Escaper interfaces should be used everywhere instead of Function<Object, String> or Function<String, String>. This will make API definition shorter and more clear and this should provide the same functionality if Formatter and Escaper are functional interfaces.

The reason they are that way is to support zero dependency mode.

See https://jstach.io/jstachio/io.jstach.jstache/io/jstach/jstache/JStacheType.html#STACHE

> [TemplateInfo::lastLoaded](https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/TemplateInfo.html#lastLoaded()) uses long as a timestamp and this should probably be java.time.Instant

Then if no time is found it would have to be nullable. Currently -1 is used for that case. Secondly File returns long:

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html#lastModified()

Normally I would agree though.

> I'm not sure what [Escaper::apply]
(https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/Escaper.html#apply) method does. it says

> But I'm not sure where the Appender comes into the picture. My understanding was that Escaper is just a pure function to escape strings... Are these implementation details that should not leak into a public API definition?

It is not a pure function. That is the point. It mutates an appendable. However an Escaper can act like a pure function. See: https://jstach.io/jstachio/io.jstach.jstache/io/jstach/jstache/JStacheContentType.html

as well as:

https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/Escaper.html#of(java.util.function.Function)

I will say I have been back and forth on the fence whether `JStacheContentType -> JStacheEscaper`.

Your original implementation IIRC was decorating an Appendable which would have required a new Escaper being created on every call. That is part of the reason why `Appender` exists. It is for performance. 

I originally didn't plan on making JStachio focused on performance but the reality is being super fast and crushing benchmarks draws attention particularly since I think good amount of folks using this library might target native (graal vm native).

`Appender` probably could be hidden I just haven't figured out how yet. My plan is to actually shrink the library for 2.0 but for now it is too hard to remove. Ditto for filter which is next:


> JStashioFilter
> I'm not really a user of jstachio, but from reading the docs I'm not sure what exactly JStashioFilter does and what JStachioFilter.FilterChain is. Should this really be a public API? It looks more like an implementation detail for me...

It basically is. Because of the ServiceLoader mechanism needing public access there wasn't a clean way to do hide this. Ultimately this is to support dynamic changes to the template. See: https://jstach.io/jstachio/io.jstach.opt.jmustache/module-summary.html

See in normal web development you keep the server running and modify the template over and over. Ideally we would run a compiler and compile the template and have a special class loader similar to how JSP and various others work to support such. However this is non trivial in part because of annotation processing. Luckily Mustache is a pretty simple language and JMustache happens to almost be equivalent to JStachio when setup correctly including even whitespace handling (which btw was the biggest pain in the ass).

In the long run I plan on rewriting the tokenizer and separating it out so that we can implement our own reflection based jstachio that mirrors all the functionality in static generated jstachio.

Anyway I have purposely kept JStachioFilter pretty vague at the moment. I suppose I could go into more detail with it but I doubt most will use it. I chose its name because its analogous to a servlet filter.

That being said it could be useful for caching or perhaps whitespace removal (ie make rendered HTML smaller).

The other option was decorating a `JStachio` instance to do fallback to JMustache but this turned out to be nontrivial to support particularly for Spring which prefers composition.

**Comment by sviperll** [comment]
> Anyway I have purposely kept JStachioFilter pretty vague at the moment.
> I suppose I could go into more detail with it but I doubt most will use it.
> I chose its name because its analogous to a servlet filter.

Maybe we should add a Javadoc comment on the class that says that it is internal API that is required for SPI implementation, so that people are less likely to rely on it for now...

**Comment by sviperll** [comment]
>> [TemplateInfo::lastLoaded](https://jstach.io/jstachio/io.jstach.jstachio/io/jstach/jstachio/TemplateInfo.html#lastLoaded()) uses long as a timestamp and this should probably be java.time.Instant
>>
> Then if no time is found it would have to be nullable. Currently -1 is used for that case. Secondly File returns long:
>
> https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html#lastModified()
>
> Normally I would agree though.

"Modern API" uses `FileTime` though: https://docs.oracle.com/javase/7/docs/api/java/nio/file/attribute/BasicFileAttributes.html#lastAccessTime()

**Reply by agentgt** [reply]
I'm embarrassed to say I was unaware of that class. I wonder what I good way to represent missing is. I assume epoch.

Much of the reason I picked the `long` was because of URL resources and other Mustache implementations using `long`.

For example: https://github.com/jknack/handlebars.java/blob/1ca4f1ba22680cc313e23ca709f24decb9990316/handlebars/src/main/java/com/github/jknack/handlebars/io/URLTemplateSource.java#L121

(albeit that is private there are other ones that I believe use `long` in public API).

I might be able to remove the whole last modified altogether.

**Comment by sviperll** [comment]
> It is not a pure function. That is the point. It mutates an appendable. 
>
> Your original implementation IIRC was decorating an Appendable which would have required a new Escaper being created on every call. That is part of the reason why Appender exists. It is for performance.
> 
> I originally didn't plan on making JStachio focused on performance but the reality is being super fast and crushing benchmarks draws attention particularly since I think good amount of folks using this library might target native (graal vm native).

maybe merge `Escaper` and `Appender`?

````
interface Appender<A extends Appendable, E extends Exception> {
    static Appender<Appendable, IOException> defaultAppender() {...}
    static Appender<StringBuilder, RuntimeException> stringAppender() {...}
    static <A extends Appendable, E extends Exception> Appender<A, E> escaping(Function<Object, String> escapeFunction, Appender<A, E> appender) {...}

    /** Each appender caches Appender instance that bypasses escaping function. */
    Appender<A, E> unescaped();

    A append(A appender, Object value) throws E;
    A append(A appender, int value) throws E;
    A append(A appender, short value) throws E;
    A append(A appender, long value) throws E;
    A append(A appender, CharSequence value) throws E;
    ...
}
````

**Reply by agentgt** [reply]
I sort of had it that way at one point (including the the parameterized exception) as your generated code would have the two appendables I would have two appenders.

The issue was that it made implementing an Escaper *look* scary (because it would have all those methods) as well as makes the Appender do more than it really should with the whole `unescaped`.

And then there is composition problems with the parameterized exception. You can't use certain Escapers with certain Appenders. And what does an Escaper that is now an appender return for `unescaped()` (I suppose itself).

One way to merge would be to repeat all the `append` methods as `escape` where `append` writes directly and `escape` does escaping. You could make it ergonomic with default methods. This is indeed what some other templating languages do like JTE do (actually JTE doesn't just have duplicate methods but wraps a writer similar to what was done in static-mustache but gets around the whole object creation by creating pools but lets ignore that for example sake).

I think I might just make `Appender` like `JStachioFilter` where I just say it is an internal implementation. It is unlikely to change anyway so I feel its pretty safe.

**Reply by agentgt** [reply]
Let me think about this some more. I think I see your point.

**Reply by agentgt** [reply]
After thinking about overnight the easiest solution for now is to make Appender `sealed`.

**Comment by agentgt** [comment]
@sviperll I have improved some of the doc on Escaper and Formatters as well as Appender.

Appender is sealed now so it really doesn't matter that it is exposed.

I know why I did not make `Escaper` and `Formatter` FunctionalInterface is because I want the ability to add another method later if I need to.

I don't think it is good idea for longevity of the API to allow:

```
Escaper e = s -> s
```

Instead they should always:

```
Escaper e = Escaper.of(s -> s);
```

As for `NullToleratingFormatter` I disagree there as well. The intent is to be compliant with the Spec and not just allow `null`. The spec could add other behavior and or I could change our default behavior. I was going to call it "StrictFormatter" as JMustache uses the term "strict" for compliance with spec but that seems misleading as our DefaultFormatter is ironically more "strict".
